### PR TITLE
Simplify Block serialization/deserialization

### PR DESCRIPTION
**Pull request by haozhun**
Previously, there are 3 sections in a serialized Block:
* Name - written by BlockEncodingSerde
  * e.g. "ARRAY" indicating it's an ArrayBlock
* Metadata - written by BlockEncodingFactory
  * e.g. "INT_ARRAY" indicating the inner Block type of an ArrayBlock
  * e.g. 4 indicating the entry size of FixedWidthBlock
* Data - written by BlockEncoding

This 3-section scheme makes things more complicated:
* Not recursive - For an ArrayBlock with IntArrayBlock inside, the encoding
  of IntArrayBlock isn't the same as a stand-alone IntArrayBlock. Specifically,
  part of it is put in the metadata section, and part of it in the data section.
  This requires each invididual encoding to handle nested blocks,
  which introduces both code duplication and complexity.
* It is harder to learn and understand the implementations with
  two interfaces (BlockEncodingFactory and BlockEncoding),
  which don't have a very clear separation of responsibility, for each Block.

This commit solves both problems:
* Only a single interface (BlockEncoding) is now needed.
* A block inside another block now serializes the exact same way as a standalone one.
  The outer block no longer need to handle any of the details of the inner one.

It also resolves the TODO item that replacementBlockForWrite was not respected for
nested Blocks.

### CODE REVIEW

**Comment by wenleix** [comment]
Relates to https://github.com/prestodb/presto/issues/10386

File: presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public String getName()
+    {
+        return NAME;
+    }
+
+    @Override
+    public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput input)
+    {
+        // We write the actual underlying block, so we will never need to read a lazy block
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java]: nit: ".. need to read a GroupIdBlock`

File: presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceOutput, Block block)
+    {
+        // We implemented replacementBlockForWrite, so we will never need to write a lazy block
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java]: ditto

File: presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+
+import java.util.Optional;
+
+public class GroupByIdBlockEncoding
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java]: Does this class have the wrong name?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        implements BlockEncoding
+{
+    public static final String NAME = "LAZY";
+
+    public GroupByIdBlockEncoding() {}
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java]: you don't need an empty constructor

File: presto-spi/src/main/java/com/facebook/presto/spi/block/LazyBlockEncoding.java [hunk 5]
```diff
@@ -21,7 +21,7 @@
 public class LazyBlockEncoding
         implements BlockEncoding
 {
-    private static final String NAME = "LAZY";
+    public static final String NAME = "LAZY";
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/LazyBlockEncoding.java]: There is already an encoding above with this name

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlockEncoding.java [hunk 6]
```diff
@@ -74,8 +63,11 @@ public void writeBlock(SliceOutput sliceOutput, Block block)
 
         int entriesStartOffset = offsets[offsetBase];
         int entriesEndOffset = offsets[offsetBase + positionCount];
-        keyBlockEncoding.writeBlock(sliceOutput, mapBlock.getKeys().getRegion(entriesStartOffset, entriesEndOffset - entriesStartOffset));
-        valueBlockEncoding.writeBlock(sliceOutput, mapBlock.getValues().getRegion(entriesStartOffset, entriesEndOffset - entriesStartOffset));
+
+        TypeSerde.writeType(sliceOutput, mapBlock.keyType);
+
+        blockEncodingSerde.writeBlock(sliceOutput, mapBlock.getKeys().getRegion(entriesStartOffset, entriesEndOffset - entriesStartOffset));
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlockEncoding.java]: maybe add a helper to mapBlock to get the key block... actually, maybe rename `getKeys` to `getRawKeysBlock`, and make this the default behavior

File: presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockEncoding.java [hunk 7]
```diff
@@ -40,24 +32,21 @@ public String getName()
...
         int[] fieldBlockOffsets = rowBlock.getFieldBlockOffsets();
-
         int startFieldBlockOffset = fieldBlockOffsets[offsetBase];
         int endFieldBlockOffset = fieldBlockOffsets[offsetBase + positionCount];
-        for (int i = 0; i < fieldBlockEncodings.length; i++) {
-            fieldBlockEncodings[i].writeBlock(sliceOutput, rowBlock.getFieldBlocks()[i].getRegion(startFieldBlockOffset, endFieldBlockOffset - startFieldBlockOffset));
+
+        sliceOutput.appendInt(numFields);
+        for (int i = 0; i < numFields; i++) {
+            blockEncodingSerde.writeBlock(sliceOutput, rowBlock.getFieldBlocks()[i].getRegion(startFieldBlockOffset, endFieldBlockOffset - startFieldBlockOffset));
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockEncoding.java]: same comment about a helper method

File: presto-spi/src/main/java/com/facebook/presto/spi/block/SingleMapBlockEncoding.java [hunk 8]
```diff
@@ -61,23 +51,30 @@ public String getName()
...
     {
         SingleMapBlock singleMapBlock = (SingleMapBlock) block;
+        TypeSerde.writeType(sliceOutput, singleMapBlock.keyType);
+
         int offset = singleMapBlock.getOffset();
         int positionCount = singleMapBlock.getPositionCount();
-        keyBlockEncoding.writeBlock(sliceOutput, singleMapBlock.getKeyBlock().getRegion(offset / 2, positionCount / 2));
-        valueBlockEncoding.writeBlock(sliceOutput, singleMapBlock.getValueBlock().getRegion(offset / 2, positionCount / 2));
+        blockEncodingSerde.writeBlock(sliceOutput, singleMapBlock.getKeyBlock().getRegion(offset / 2, positionCount / 2));
+        blockEncodingSerde.writeBlock(sliceOutput, singleMapBlock.getValueBlock().getRegion(offset / 2, positionCount / 2));
```
Comment (dain) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/SingleMapBlockEncoding.java]: a helper here would be a big help

File: presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java (hunk 1 above)
Comment (haozhun) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java]: The class is not used. I deleted it.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java (hunk 2 above)
Comment (haozhun) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java]: The class is not used. I deleted it.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java (hunk 3 above)
Comment (haozhun) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java]: The class is not used. Deleting it.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java (hunk 4 above)
Comment (haozhun) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/GroupByIdBlockEncoding.java]: The class is not used. Deleting it.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/LazyBlockEncoding.java (hunk 5 above)
Comment (haozhun) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/LazyBlockEncoding.java]: `GroupByIdBlockEncoding` is not used. Deleting it.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlockEncoding.java (hunk 6 above)
Comment (haozhun) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlockEncoding.java]: Done renaming. I will try out introducing the helper separately.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockEncoding.java (hunk 7 above)
Comment (haozhun) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockEncoding.java]: Done renaming. I will try out introducing the helper separately.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/SingleMapBlockEncoding.java (hunk 8 above)
Comment (haozhun) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/SingleMapBlockEncoding.java]: Done renaming. I will try out introducing the helper separately.

**Pull request merged**